### PR TITLE
fix(issue-alerts): paginate over groups instead of returning graph bu…

### PR DESCRIPTION
**Pull request by Zandew**
Preview endpoint now returns the groups that would have triggered the rule instead of buckets for a graph. Looks like [this](https://github.com/getsentry/sentry/pull/40338) on the frontend.

### CODE REVIEW

File: src/sentry/api/endpoints/project_rule_preview.py [hunk 1]
```diff
@@ -51,4 +50,17 @@ def post(self, request: Request, project) -> Response:
...
+        response = self.paginate(
+            request=request,
+            queryset=results,
+            order_by="-id",
+            on_results=lambda x: serialize(x, request.user),
+        )
+
+        response.data = {
+            "data": response.data,
+            "totalCount": len(results),
```
Comment (rjo100) [inline src/sentry/api/endpoints/project_rule_preview.py:63]: change to `results.count()`. much more efficient

File: tests/sentry/rules/history/test_preview.py [hunk 2]
```diff
@@ -4,64 +4,64 @@
...
-            assert result[i].count == (1 if i % 5 else 0)
+        result = preview(self.project, conditions, [], "all", "all", 120)
+        assert len(result) == result2
 
     def test_first_seen(self):
-        hours = self._set_up_first_seen()
+        self._set_up_first_seen()
         self._test_preview(
-            hours, "sentry.rules.conditions.first_seen_event.FirstSeenEventCondition"
+            "sentry.rules.conditions.first_seen_event.FirstSeenEventCondition", 670, 134
```
Comment (rjo100) [inline tests/sentry/rules/history/test_preview.py:50]: can we make these some sort of static variable that's linked up with `i` and `j` somehow? bare integers can be hard to parse/understand

Comment (Zandew) [reply tests/sentry/rules/history/test_preview.py]: tried to make some sense of it, but i think the numbers are pretty unintuitive to begin with

Comment (rjo100) [reply tests/sentry/rules/history/test_preview.py]: no that's perfect. just like we expect them to move in proportion

File: src/sentry/api/endpoints/project_rule_preview.py [hunk 3]
```diff
@@ -51,4 +50,17 @@ def post(self, request: Request, project) -> Response:
...
+            request=request,
+            queryset=results,
+            order_by="-id",
+            on_results=lambda x: serialize(x, request.user),
+        )
+
+        response.data = {
+            "data": response.data,
+            "totalCount": results.count(),
+        }
```
[the remark this answers is not in the document]
Comment (Zandew) [reply src/sentry/api/endpoints/project_rule_preview.py]: looks like not all paginators support `count_hits`

**Review by rjo100 — APPROVED** [review]

File: src/sentry/api/base.py [hunk 4]
```diff
@@ -364,7 +365,12 @@ def paginate(
                 sentry_sdk.set_tag(
                     "query.per_page.grouped", format_grouped_length(per_page, [1, 10, 50, 100])
                 )
-                cursor_result = paginator.get_result(limit=per_page, cursor=input_cursor)
+                if isinstance(paginator, Paginator):
+                    cursor_result = paginator.get_result(
+                        limit=per_page, cursor=input_cursor, count_hits=count_hits
+                    )
+                else:
+                    cursor_result = paginator.get_result(limit=per_page, cursor=input_cursor)
```
Comment (wedamija) [inline src/sentry/api/base.py:373]: I'd say that it might be fine to let a paginator that doesn't support `count_hits` just fail here. The failure should be caught in tests, so that the developer is aware of the problem. This check here will silently ignore the problem, so they might not be aware that they're not getting hits in the end.

Comment (Zandew) [reply src/sentry/api/base.py:373]: got it, I'll make another pr to change this
